### PR TITLE
Fix for #5034

### PR DESCRIPTION
**Pull request by plato79**
This fixes the errors when installing, although environment variable to skip download still should be used like this:
`setenv PUPPETEER_SKIP_DOWNLOAD true`

### CODE REVIEW

**Comment by plato79** [comment]
@googlebot I signed it!

**Comment by plato79** [comment]
Could somebody check this? This fixes https://github.com/puppeteer/puppeteer/issues/5034 . Although it stucks for something not related with this it seems.

**Comment by 0mp** [comment]
It would be great if we could have it merged. @plato79, do you still have time to resolve the conflicts perhaps?

**Comment by jschfflr** [comment]
@mathiasbynens Not sure if supporting `freebsd` is something we want to officially commit to doing - could you take a look and resolve this accordingly?

**Comment by 0mp** [comment]
> @mathiasbynens Not sure if supporting `freebsd` is something we want to officially commit to doing - could you take a look and resolve this accordingly?

Thanks for working on this PR. It's much appreciated. 😃

File: src/BrowserFetcher.js [hunk 1]
```diff
@@ -222,6 +228,8 @@ class BrowserFetcher {
         executablePath = path.join(folderPath, archiveName(this._product, this._platform, revision), 'Chromium.app', 'Contents', 'MacOS', 'Chromium');
       else if (this._platform === 'linux')
         executablePath = path.join(folderPath, archiveName(this._product, this._platform, revision), 'chrome');
+      else if (this._platform === 'freebsd')
```
Comment (mathiasbynens) [inline src/BrowserFetcher.js:231]: Can we merge this with the `else if (this._platform === 'linux')` above?

File: src/BrowserFetcher.js [hunk 2]
```diff
@@ -67,6 +69,8 @@ function archiveName(product, platform, revision) {
       return 'chrome-linux';
     if (platform === 'mac')
       return 'chrome-mac';
+    if (platform === 'freebsd')
```
Comment (mathiasbynens) [inline src/BrowserFetcher.js:72]: Can we merge this with the `if (platform === 'linux')` above?

File: src/BrowserFetcher.js [hunk 3]
```diff
@@ -231,6 +239,8 @@ class BrowserFetcher {
         executablePath = path.join(folderPath, 'Firefox Nightly.app', 'Contents', 'MacOS', 'firefox');
       else if (this._platform === 'linux')
         executablePath = path.join(folderPath, 'firefox', 'firefox');
+      else if (this._platform === 'freebsd')
```
Comment (mathiasbynens) [inline src/BrowserFetcher.js:242]: Merge this with the check above

File: src/BrowserFetcher.js [hunk 4]
```diff
@@ -32,12 +32,14 @@ const getProxyForUrl = require('proxy-from-env').getProxyForUrl;
 const downloadURLs = {
   chrome: {
     linux: '%s/chromium-browser-snapshots/Linux_x64/%d/%s.zip',
+    freebsd: '%s/chromium-browser-snapshots/Linux_x64/%d/%s.zip',
```
Comment (mathiasbynens) [inline src/BrowserFetcher.js:35]: This and the Firefox URL below are the same ones as for Linux? In that case let's simplify the patch by removing this line and line 42 and treating `'freebsd'` as equivalent to `'linux'`.

That way this change also doesn't impact the public `puppeteer.createBrowserFetcher` API.

**Review by mathiasbynens — CHANGES_REQUESTED** [review]
I believe this patch affects the public `puppeteer.createBrowserFetcher` API and its `platform` option but it doesn't document the changes.
However, it's better not to change this API and stick to the current list of user-exposed `platform`s (`"linux"|"mac"|"win32"|"win64"`) since those are the ones we test and provide support for.

It looks like this patch is just using the Linux URLs and codepaths for `freebsd` anyway — so can we just treat `freebsd` as an alias to `linux`? That would avoid these issues.

**Comment by plato79** [comment]
> I believe this patch affects the public `puppeteer.createBrowserFetcher` API and its `platform` option but it doesn't document the changes.
> However, it's better not to change this API and stick to the current list of user-exposed `platform`s (`"linux"|"mac"|"win32"|"win64"`) since those are the ones we test and provide support for.
> 
> It looks like this patch is just using the Linux URLs and codepaths for `freebsd` anyway — so can we just treat `freebsd` as an alias to `linux`? That would avoid these issues.

The thing is puppeteer checks for the system and if it's freebsd it just doesn't work at all.. The links doesn't matter because chromium should be installed manually anyway on freebsd ( there is no official support ). You could also use dummy download links or no download links at all.. So, as I said in the first post you need to use `setenv PUPPETEER_SKIP_DOWNLOAD true` first to avoid downloading part.

**Comment by jschfflr** [comment]
Not sure if I understand you correctly: For FreeBSD, the normal linux build of Chromium doesn't work. That's why you `setenv PUPPETEER_SKIP_DOWNLOAD true` to prevent Puppeteer from downloading something and you instead install it manually. In which situation would the `BrowserFetcher` be used then?

**Comment by plato79** [comment]
> Not sure if I understand you correctly: For FreeBSD, the normal linux build of Chromium doesn't work. That's why you `setenv PUPPETEER_SKIP_DOWNLOAD true` to prevent Puppeteer from downloading something and you instead install it manually. In which situation would the `BrowserFetcher` be used then?

Yes, normal linux build doesn't work but there are custom FreeBSD builds of chromium exists which you need to compile yourself.. The thing is puppeteer still searches for freebsd in BrowserFetcher if the system is FreeBSD. This is just a dummy selection to prevent runtime errors.

**Comment by jschfflr** [comment]
@plato79 My expectation would be that we don't ever get into a situation where we'd run `BrowserFetcher` if `PUPPETEER_SKIP_DOWNLOAD` is set. Could you maybe share a quick snippet to reproduce the problem?

**Comment by jschfflr** [comment]
@plato79 Gentle ping :)

**Comment by jschfflr** [comment]
@plato79 I'm going to close this for now because it's not actionable for me but please feel free to reopen if you found out why the browser fetcher is still invoked even though `PUPPETEER_SKIP_DOWNLOAD ` was set!

**Pull request closed without merging**